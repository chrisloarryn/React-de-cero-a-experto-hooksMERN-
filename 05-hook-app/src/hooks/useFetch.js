import { useState, useEffect, useRef } from 'react'

export const useFetch = url => {
  const isMounted = useRef(true)
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null
  })
  useEffect(() => {
    return () => {
      isMounted.current = false
    }
  }, [])
  useEffect(() => {
    setState({ data: null, loading: true, error: null })
    fetch(url)
      .then(resp => resp.json())
      .then(data => {
        setTimeout(() => {
          if (isMounted.current)
            setState({
              loading: false,
              error: null,
              data
            })
          else console.log('setState() did not called')
        }, 1000)
      })
      .catch(() => {
        setState({
          data: null,
          loading: false,
          error: 'data has not been charged'
        })
      })
  }, [url])

  return state
}
