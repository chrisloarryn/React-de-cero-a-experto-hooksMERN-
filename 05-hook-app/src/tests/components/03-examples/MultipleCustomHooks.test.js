import React from 'react'
import { shallow } from 'enzyme'
import { MultipleCustomsHooks } from 'components/03-examples/MultipleCustomsHooks'
import { useFetch } from 'hooks/useFetch'
import { useCounter } from 'hooks/useCounter'

jest.mock('hooks/useFetch')
jest.mock('hooks/useCounter')

describe('Tests in <MultipleCustomHooks />', () => {
  useCounter.mockReturnValue({
    counter: 1,
    increment: () => {}
  })

  test('Should be rendered perfectly', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: false,
      error: null
    })

    const wrapper = shallow(<MultipleCustomsHooks />)
    expect(wrapper).toMatchSnapshot()
  })

  test('Should return author data', () => {
    useFetch.mockReturnValue({
      data: [{ author: 'Cristobalite', quote: 'Hello world!' }],
      loading: false,
      error: null
    })
    const wrapper = shallow(<MultipleCustomsHooks />)

    expect(wrapper.find('.alert').exists()).toBe(false)
    expect(wrapper.find('.mb-0').text().trim()).toBe('Hello world!')
    expect(wrapper.find('footer').text().trim()).toBe('Cristobalite')

    console.log(wrapper.html())
  })
})
