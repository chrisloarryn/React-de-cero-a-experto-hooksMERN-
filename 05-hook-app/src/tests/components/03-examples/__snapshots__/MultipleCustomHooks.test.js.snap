// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <MultipleCustomHooks /> Should be rendered perfectly 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    />
    <footer
      className="blockquote-footer"
    />
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
